File: sort/bubble_sort/bubble_sort.py
# Swap if arr[0]>arr[1] and Swap if arr[1]>arr[2]... -> ...Swap if arr[-3]>arr[-2] -> so on...
# arr[-1] is sorted -> arr[-2:] is sorted -> so on...


def bubble_sort(l):  # impure
    n = len(l)

    for i in range(n - 1):
        already_sorted = True
        for j in range(n - i - 1):
            if l[j] > l[j + 1]:
                l[j], l[j + 1] = l[j + 1], l[j]
                already_sorted = False
        if already_sorted:
            break  # If there were no swaps during the last iteration,the array is already sorted, and you can terminate


l = [float(i) for i in input("Enter the list: ").strip().split()]
bubble_sort(l)
print(l)

File: sort/bubble_sort/bubble_sort.cpp
#include <iostream>
#include <vector>

// n,n^2

using namespace std;

void bubble_sort(vector<float>::iterator &begin, vector<float>::iterator &end)
{
    bool already_sorted;
    for (int i = 0; i != distance(begin, end) - 1; i++)
    {
        already_sorted = true;
        for (vector<float>::iterator it = begin; it != end - i - 1; it++)
            if (*it > *(it + 1))
            {
                // float temp = *it;
                // *it = *(it + 1);
                // *(it + 1) = temp;
                swap(*it, *(it + 1));
                already_sorted = false;
            }
        if (already_sorted)
            break;
    }
}

int main()
{
    int n;

    cout << "Enter the length of the vector: ";
    cin >> n;
    vector<float> v(n, 0);

    cout << "Enter the elements: ";
    for (float &x : v)
        cin >> x;

    vector<float>::iterator begin = v.begin(), end = v.end();

    bubble_sort(begin, end); // begin, end must be variables (not literals)

    for (float x : v)
        cout << x << endl;

    return 0;
}